src/vs/editor/contrib/stickyScroll/browser/stickyScrollModelProvider.ts

```diff
@@ -0,0 +1,314 @@
+
+		return providerModelPromise.then(providerModel => {
+			if (!this.isModelValid(providerModel)) {
+				return this._invalid();
+
+			} else if (providerModelPromise === this._providerModelPromise && this._updateDebounceInfo && this._stopWatch && this._updateScheduler) {
+				const newValue = this._updateDebounceInfo.update(textModel, this._stopWatch.elapsed());
+				this._updateScheduler.defaultDelay = newValue;
+			}
+			if (token.isCancellationRequested) {
+				return Status.CANCELED;
+			}
+			return this.createStickyModel(textModel, modelVersionId, token, providerModel);
+		}).then(undefined, (err) => {
+			onUnexpectedError(err);
+			return Status.CANCELED;
+		});
+	}
+
+	/**
+	 * Method which checks whether the model returned by the provider is valid and can be used to compute a sticky model.
+	 * This method by default returns true.
+	 * @param model model returned by the provider
+	 * @returns boolean indicating whether the model is valid
+	 */
+	protected isModelValid(model: any): boolean {
+		return true;
+	}
+
+	/**
+	 * Method which checks whether the provider is valid before applying it to find the provider model.
+	 * This method by default returns true.
+	 * @param textModel text-model of the editor
+	 * @returns boolean indicating whether the provider is valid
+	 */
+	protected isProviderValid(textModel: ITextModel): boolean {
+		return true;
+	}
+
+	/**
+	 * Abstract method which creates the model from the provider
+	 * @param textModel	text-model of the editor
+	 * @param textModelVersionId text-model version ID
+	 * @param token cancellation token
+	 */
+	protected abstract createModelFromProvider(textModel: ITextModel, textModelVersionId: number, token: CancellationToken): any;
+
+	/**
+	 * Abstract method which computes the sticky model from the model returned by the provider
+	 * @param textModel text-model of the editor
+	 * @param textModelVersionId text-model version ID
+	 * @param token cancellation token
+	 * @param model model returned by the provider
+	 */
+	protected abstract createStickyModel(textModel: ITextModel, textModelVersionId: number, token: CancellationToken, model: any): Status;
+}
+
+class StickyModelFromCandidateOutlineProvider extends StickyModelCandidateProvider {
+
+	constructor(@ILanguageFeaturesService private readonly _languageFeaturesService: ILanguageFeaturesService) {
+		super();
+	}
+
+	protected createModelFromProvider(textModel: ITextModel, modelVersionId: number, token: CancellationToken): Promise<OutlineModel> {
+		return OutlineModel.create(this._languageFeaturesService.documentSymbolProvider, textModel, token);
+	}
+
+	protected createStickyModel(textModel: TextModel, modelVersionId: number, token: CancellationToken, model: OutlineModel): Status {
+		const { stickyOutlineElement, providerID } = StickyElement.fromOutlineModel(model, this._stickyModel?.outlineProviderId);
```

Comment:
The `StickyElement.fromOutlineModel` could be moved into this class (`StickyModelFromCandidateOutlineProvider`)